crates/containerd-shim-wasm/src/sandbox/exec.rs

```diff
@@ -0,0 +1,252 @@
+            pid: pid.as_raw() as u32,
+            status: status as u32,
+        })),
+        WaitStatus::Signaled(pid, sig, dumped) => {
+            info!("child {} killed by signal {}, dumped: {}", pid, sig, dumped);
+            Ok(Status::Exited(ExitStatus {
+                pid: pid.as_raw() as u32,
+                status: 128 + sig as u32,
+            }))
+        }
+        WaitStatus::StillAlive => Ok(Status::Running),
+        _ => Err(Error::Others(format!("unexpected wait status: {:?}", info))),
+    }
+}
+
+impl PidFD {
+    // wait for the process referred to by the pidfd to exit
+    //
+    // If you want more control over waiting you can use `as_raw_fd()` and call `waitid` yourself.
+    pub fn wait(&self) -> Result<ExitStatus, Error> {
+        let ws = wait(self.fd, false)?;
+        Ok(ws.into())
+    }
+
+    pub fn wait_no_hang(&self) -> Result<Status, Error> {
+        wait(self.fd, true)
+    }
+
+    // Send the specified signal to the process referred to by the pidfd.
+    pub fn kill(&self, sig: i32) -> Result<(), Error> {
+        let ret = unsafe { pidfd_send_signal(self.fd, sig, ptr::null_mut(), 0) };
+        if ret == -1 {
+            return Err(std::io::Error::last_os_error().into());
+        }
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+pub struct ExitStatus {
+    pub pid: u32,
+    pub status: u32,
+}
+
+pub enum Context {
+    // Parent stores the pid of the child process and the pidfd that can be used to, for instance, wait on the child.
+    Parent(u32, PidFD),
+    Child,
+}
+
+pub fn has_cap_sys_admin() -> bool {
+    let caps = caps::read(None, CapSet::Effective).unwrap();
+    caps.contains(&Capability::CAP_SYS_ADMIN)
+}
+
+// This is is a wrapper around clone3 which allows us to pass a pidfd
+// This works otherwise just like normal forking semantics:
+// If this is the parent, the result will be Ok(Context::Parent(pid, pidfd)), where the pid is the pid of the new process.
+// If this is the child, the result will be Ok(Context::Child).
+//
+// Code that runs in the child must not do things like access locks or other shared state.
+//
+// Optionally you can pass in a reference to a file descriptor which will be populated with the pidfd (see pidfd_open(2)).
+pub unsafe fn fork(cgroup: Option<&Box<dyn Cgroup>>) -> Result<Context, Error> {
+    let mut builder = Clone3::default();
+
+    let mut fd: RawFD = -1;
+    builder.flag_pidfd(&mut fd);
+
+    builder.exit_signal(SIGCHLD as u64).flag_ptrace();
+
+    let is_root = has_cap_sys_admin();
+
+    let mut cgfd: RawFD = -1; // Keep the fd alive until we return
+    if is_root {
+        if let Some(cgroup) = &cgroup {
+            if cgroup.version() == CgroupVersion::V2 {
+                cgfd = cgroup.open()?;
+                builder.flag_into_cgroup(&cgfd).flag_newcgroup();
+            }
+        }
+    } else {
+        debug!("no CAP_SYS_ADMIN, not creating new namespaces");
+    }
+
+    let res = { builder.call() };
```

Comment:
This is the unsafe part, perhaps since we are not really giving a way to mis-use it or do bad things with memory we don't need to be unsafe except for just this call?

Reply:
Yep~ We can just wrap this individual call with `unsafe`.